Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java

```diff
@@ -1,125 +1,147 @@
+                String json = new GsonBuilder().create().toJson(captureStarted);
+                long ts = System.currentTimeMillis();
+                conn.outputWriter.write(json);
+                conn.outputWriter.flush();
+                return ts;
+              }
+            });
   }
 
-  public ArrayList<Long> downloadServerTimeStamps() {
-    if (serverSocketPort == null || !socket.isConnected()) {
-      Log.d(ContentValues.TAG, "Can't find server to download timestamps. No port bound or connection is dead.");
-      return null;
-    }
-    DownloadServerLogs downloadServerLogsTask = new DownloadServerLogs();
-    try {
-      return downloadServerLogsTask.execute().get();
-    } catch (ExecutionException | InterruptedException e) {
-      e.printStackTrace();
-    }
-    return null;
+  @RequiresApi(api = VERSION_CODES.N)
+  public void sendOcrResults(
+      ArrayList<Text> resultsOcr, long recordingStartMillis, int fpsRecording) {
+    connection.thenApply(
+        conn -> {
+          synchronized (conn) {
+            ArrayList<FrameData> framesMetaData = parseTextObjects(resultsOcr);
+            try {
+              HostData hostData = new HostData(framesMetaData, recordingStartMillis, fpsRecording, hostSyncTimestamp.get());
+              ServerAction receiveResults = new ServerAction(2, "Sending results");
+              conn.outputWriter.write(new GsonBuilder().create().toJson(receiveResults));
+              conn.outputWriter.flush();
+
+              conn.outputWriter.write(new GsonBuilder().create().toJson(hostData));
+              conn.outputWriter.flush();
+            } catch (ExecutionException | InterruptedException e) {
+              e.printStackTrace();
+            }
+          }
+          return null;
+        });
   }
 
-  //TODO: Replace AsyncTask with something more robust
-  /** Requests the server to send the timestamps of key-presses. */
-  private class DownloadServerLogs extends AsyncTask<Void, Void, ArrayList<Long>> {
-    @Override
-    protected ArrayList<Long> doInBackground(Void... values) {
-      if (serverSocketPort == null) {
-        Log.d(
-            ContentValues.TAG,
-            "No server to read input from. Check if server communication is working.");
-      }
-      ArrayList<Long> serverTimestamps = new ArrayList<>();
-      try {
-        Log.d(ContentValues.TAG, "Sending request to download timestamps");
-        outputWriter.write("send timestamps" + "*");
-        outputWriter.flush();
-        Log.d(ContentValues.TAG, "Request sent to server. Waiting to receive timestamps..");
-        String message = inputReader.readLine();
-        serverStartTimestamp = Long.parseLong(message);
-
-        while (message != null) {
-          serverTimestamps.add(Long.valueOf(message));
-          Log.d(ContentValues.TAG, "Server TimeStamp:" + message);
-          message = inputReader.readLine();
+  private ArrayList<FrameData> parseTextObjects(ArrayList<Text> resultsOcr) {
```

Comment:
It would be better to make FrameData a public class and do this conversion in the class that handles the OCR.  That way you do not have any dependency from this class on the mlkit library, which makes it easier to change or add additional approaches if we ever need to do that.  It also makes it easier to test this class, as we don't need to be able to create an mlkit Text object.